### PR TITLE
Make ValuesSourceConfig behave like a config object

### PR DESCRIPTION
**Pull request by not-napoleon**
This PR has two major changes for how ValuesSourceConfig works.  First, it makes VSConfig immutable.  It had been effectively immutable already, in that nothing outside of its factory methods called its mutators, but they still existed.  Now, it's obvious at a glance that the class is immutable, which should make it easier to reason about, and to pass around without worry.

A consequence of that change, a lot of places where we previously picked out parts of VSConfig to pass into aggregators, we now just pass the whole config in.  This simplifies constructor arguments, lets us merge some supplier interfaces together, and makes it easier for future developers to add aggregators that might need some part of the config we hadn't used before - it no longer requires changing a bunch of signatures up the call chain to add a new parameter.

The second major change is that VSConfig no longer creates null values sources.  Previously, we'd used a null ValuesSource to communicate that there were no values (aka the unmapped-and-no-missing case).  The aggregation builder would detect this and call `createUnmapped` instead of `doCreateInternal` on the aggregator factory.  But since the field was nullable, a lot of aggregators have to carry around support for getting passed a null values source, even in cases where they never would be.  Other places, we effectively have a parallel aggregator in the same class for nulls.  In a follow-up PR, we can remove all of that needless null handling and just create proper, optimized, empty aggregators for the `createUnmapped` path to return.

Besides these major changes, the rest of this PR plumbs through the low hanging fruit of passing the VSConfig directly into aggregators.  Not all aggregators were converted, because some of the more complicated ones I'd rather do in a smaller PR.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java [hunk 1]
```diff
@@ -269,8 +275,36 @@ public ValuesSourceConfig(
...
+        this.format = format == null ? DocValueFormat.RAW : format;
         this.nowSupplier = nowSupplier;
 
+        if (!valid()) {
+            // TODO: resolve no longer generates invalid configs.  Once VSConfig is immutable, we can drop this check
+            throw new IllegalStateException(
+                "value source config is invalid; must have either a field context or a script or marked as unwrapped");
+        }
+
+        final ValuesSource vs;
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java:289]: This'd probably be a little cleaner as a `private final` method with early returns.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java [hunk 2]
```diff
@@ -285,8 +319,12 @@ public FieldContext fieldContext() {
...
     }
 
-    public boolean unmapped() {
-        return unmapped;
+    /**
+     * Returns true if the values source configured by this object can yield values.  We might not be able to yield values if, for example,
+     * the specified field does not exist on this index.
+     */
+    public boolean hasValues() {
+        return fieldContext != null || script != null || missing != null;
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java:327]: Isn't `|=` the compound bitwise-or-then-assign operator?

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java [hunk 3]
```diff
@@ -154,10 +154,9 @@ private static ValuesSourceConfig internalResolve(QueryShardContext context,
         if (valuesSourceType == null) {
             valuesSourceType = defaultValueSourceType;
         }
-        config = new ValuesSourceConfig(valuesSourceType, fieldContext, unmapped, aggregationScript, scriptValueType, context::nowInMillis);
-        config.format(resolveFormat(format, valuesSourceType, timeZone, fieldType));
-        config.missing(missing);
-        config.timezone(timeZone);
+        DocValueFormat docValueFormat = resolveFormat(format, valuesSourceType, timeZone, fieldType);
+        config = new ValuesSourceConfig(valuesSourceType, fieldContext, unmapped, aggregationScript, scriptValueType, missing,
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java:158]: Run it through the code-formatter?

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java [hunk 4]
```diff
@@ -239,18 +241,19 @@ public static ValuesSourceConfig resolveFieldOnly(MappedFieldType fieldType,
...
-    private LongSupplier nowSupplier;
+    private final FieldContext fieldContext;
+    private final AggregationScript.LeafFactory script;
+    private final ValueType scriptValueType;
+    private final boolean unmapped;
+    private final DocValueFormat format;
+    private final Object missing;
+    private final ZoneId timeZone;
+    private final LongSupplier nowSupplier;
+    private ValuesSource valuesSource;
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java:256]: Can this be final too?

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java (hunk 2 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java:327]: Ignore this. On my browser `|=` and `!=` look very similar.

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/SpatialPluginTests.java [hunk 5]
```diff
@@ -35,8 +35,8 @@ public void testGeoCentroidLicenseCheck() {
             List<Consumer<ValuesSourceRegistry.Builder>> registrar = plugin.getAggregationExtentions();
             registrar.forEach(c -> c.accept(registryBuilder));
             ValuesSourceRegistry registry = registryBuilder.build();
-            GeoCentroidAggregatorSupplier centroidSupplier = (GeoCentroidAggregatorSupplier) registry.getAggregator(
-                new ValuesSourceConfig(GeoShapeValuesSourceType.instance(), null, false, null, null, null),
+            MetricAggregatorSupplier centroidSupplier = (MetricAggregatorSupplier) registry.getAggregator(
+                new ValuesSourceConfig(GeoShapeValuesSourceType.instance(), null, true, null, null, null, null, null, null),
```
Comment (nik9000) [inline x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/SpatialPluginTests.java:39]: This is a super long number of arguments! I'm glad we don't call the ctor very much though.

**Review by nik9000 — APPROVED** [review]
Left some small stuff but LGTM so long as CI likes it.

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/SpatialPluginTests.java (hunk 5 above)
Comment (not-napoleon) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/SpatialPluginTests.java:39]: Yeah, outside of testing, the only thing that should be calling that constructor is the various resolve factory methods, which are what the rest of the framework should call.  Not that they don't have a lot of arguments too.  I suppose we could have a default constructor for tests, but I don't like adding production code that's only supposed to be called from tests.

Comment (nik9000) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/SpatialPluginTests.java:39]: How often do we call it in tests? It doesn't look like we call it directly very often. I think mostly we use the production resolution framework.

Comment (not-napoleon) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/SpatialPluginTests.java:39]: Correct, there's only a handful of cases we call the constructor from tests.  I meant I didn't want to add a new test-only constructor just to clean up this case.

Comment (nik9000) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/SpatialPluginTests.java:39]: I'm happy with that, yeah.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java [hunk 6]
```diff
@@ -106,17 +105,16 @@ protected XPackLicenseState getLicenseState() {
...
-                -> new GeoShapeBoundsAggregator(name, aggregationContext, parent, (GeoShapeValuesSource) valuesSource,
-                wrapLongitude, metadata));
+            (GeoBoundsAggregatorSupplier) (name, aggregationContext, parent, valuesSourceConfig, wrapLongitude, metadata)
+                -> new GeoShapeBoundsAggregator(name, aggregationContext, parent, valuesSourceConfig, wrapLongitude, metadata));
     }
 
     private void registerGeoShapeCentroidAggregator(ValuesSourceRegistry.Builder builder) {
         builder.register(GeoCentroidAggregationBuilder.NAME, GeoShapeValuesSourceType.instance(),
-            (GeoCentroidAggregatorSupplier) (name, aggregationContext, parent, valuesSource, metadata)
+            (MetricAggregatorSupplier) (name, valuesSourceConfig, aggregationContext, parent, metadata)
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java:113]: If we only do it once I guess it doesn't buy much. I think it'd be a little clearer that we're *just* checking the license and then doing the ctor reference thing, but its no big deal.

**Pull request merged**